### PR TITLE
Initial scripts to create and tear down test resources

### PR DESCRIPTION
**Pull request by heaths**
These are the initial scripts to set up and tear down test resources based on a convention we decided. These were testing on Windows and linux using bash, cmd, and pwsh.

### CODE REVIEW

**Comment by pakrym** [comment]
Can we enable dot-sourcing the script so environment variables are set automatically in current powershell terminal?

```
. .\eng\New-TestResource.ps1 ...
```

File: eng/New-TestResources.ps1 [hunk 1]
```diff
@@ -0,0 +1,164 @@
...
+    $templateParameters.Add('provisionerClientSecret', $ProvisionerClientSecret)
+}
+if ($TestApplicationOid) {
+    $templateParameters.Add('testApplicationOid', $TestApplicationOid)
+}
+if ($AdditionalParameters) {
+    $templateParameters += $AdditionalParameters
+}
+
+# Try to detect the shell based on the parent process name (e.g. launch via shebang).
```
Comment (pakrym) [inline eng/New-TestResources.ps1]: Fancy! 👍

**Review by pakrym — APPROVED** [review]

**Comment by heaths** [comment]
> Can we enable dot-sourcing the script so environment variables are set automatically in current powershell terminal?
> 
> ```
> . .\eng\New-TestResource.ps1 ...
> ```

Because setting the environment variables in the current session won't persist, though I suppose that doesn't preclude doing both. But the idea we discussed was to save these to a file anyway. Let's sync back up on that and determine if we're ready for that given the other work going on with credentials. If not, I agree doing both would be great.

That said, I still wouldn't allow for it to be sourced. I can just as easily set environment variables in teh script. Sourcing it makes all variables and functions global, and you have to declare the body as a function itself, nor could you support bootstrapping PowerShell from cmd or bash (or other shells) like we plan to (next iteration). In general, you shouldn't source anything other than $profile. If you want to allow people to import into the global namespace (or any other declared namespace), use modules, which can be unloaded.

File: eng/New-TestResources.ps1 [hunk 2]
```diff
@@ -0,0 +1,305 @@
...
+Get-ChildItem -Path $root -Filter $templateFileName -Recurse | ForEach-Object {
+    $templateFile = $_.FullName
+
+    Write-Verbose "Found template '$templateFile'"
+    $templateFiles += $templateFile
+}
+
+if (!$templateFiles) {
+    Write-Warning -Message "No template files found under '$root'"
+    exit
```
Comment (pakrym) [inline eng/New-TestResources.ps1]: exit 1?

File: eng/New-TestResources.ps1 [hunk 3]
```diff
@@ -0,0 +1,305 @@
...
+        $script:TestApplicationOid = $testServicePrincipal.Id
+    }
+}
+
+# Format the resource group name based on resource group naming recommendations and limitations.
+$resourceGroupName = "rg-{0}-$baseName" -f ($ServiceDirectory -replace '[\\\/]', '-').Substring(0, [Math]::Min($ServiceDirectory.Length, 90 - $BaseName.Length - 4)).Trim('-')
+
+# Tag the resource group to be deleted after a certain number of hours if specified.
+$tags = @{
+    ServiceDirectory = $ServiceDirectory
```
Comment (pakrym) [inline eng/New-TestResources.ps1]: What about adding a $env:USERNAME as a 'Creator' tag?

File: eng/New-TestResources.ps1 [hunk 4]
```diff
@@ -0,0 +1,305 @@
...
+
+    $templateFileParameters = $templateParameters.Clone()
+    foreach ($key in $templateParameters.Keys) {
+        if ($templateParameterNames -notcontains $key) {
+            Write-Verbose "Removing unnecessary parameter '$key'"
+            $templateFileParameters.Remove($key)
+        }
+    }
+
+    $preDeploymentScript = $templateFile | Split-Path | Join-Path -ChildPath 'test-resources-pre-deploy.ps1'
```
Comment (pakrym) [inline eng/New-TestResources.ps1]: I think the name is fine. Did you consider dropping `-deploy` suffix? `test-resources-pre/post`

**Review by pakrym — APPROVED** [review]

**Comment by pakrym** [comment]
I like it. Very clean!

File: eng/New-TestResources.ps1 [hunk 5]
```diff
@@ -0,0 +1,305 @@
...
+
+        # Work around bug that makes the first few characters of environment variables be lowercase.
+        $key = $key.ToUpperInvariant()
+
+        if ($variable.Type -eq 'String' -or $variable.Type -eq 'SecureString') {
+            $deploymentOutputs[$key] = $variable.Value
+
+            if ($env:SYSTEM_TEAMPROJECTID) {
+                # Running in Azure Pipelines.
+                Write-Host "##vso[task.setvariable variable=$key;issecret=true;]$($variable.Value)"
```
Comment (pakrym) [inline eng/New-TestResources.ps1]: It's slightly unfortunate that all variables would become secrets, having some visible might benefit debugging. We can fix this later.

File: eng/New-TestResources.ps1 (hunk 2 above)
Comment (heaths) [reply eng/New-TestResources.ps1]: That would be an error and, by default, fail the task. The idea here is that the script is always executed and will react accordingly.

File: eng/New-TestResources.ps1 (hunk 4 above)
Comment (heaths) [reply eng/New-TestResources.ps1]: That was actually my first names. :) I was afraid they were 1) non-descript, and 2) hard to expand on should we want to add similar stuff later. But that's probably very unlikely.

File: eng/New-TestResources.ps1 (hunk 5 above)
Comment (heaths) [reply eng/New-TestResources.ps1]: There's no good way to differentiate, unfortunately. We have to change output SecureStrings to Strings to even get the value, so everything coming to us here would be String anyway (in practice) which, I agree, would be good to switch on for the value of `issecret`.

I will add a comment, though, explaining this / why and maybe we can change it in the future if we can think of a better way.

**Pull request merged**